### PR TITLE
feat: scroll execution strategy

### LINKED ISSUE #5: Scroll `EvmExecutor` implementation
### Describe the feature

# Overview
To execute blocks using [`scroll-revm`](https://github.com/frisitano/scroll-revm) we must implement an `EvmExecutor` (and associated assets) that wraps `scroll-revm` and executes blocks.  An example of `optimism` specific `Executor` assets can be found [here](https://github.com/paradigmxyz/reth/blob/main/crates/optimism/evm/src/execute.rs).  Logic from the [`stateless-block-verifier`](https://github.com/scroll-tech/stateless-block-verifier/tree/master) can be used as a guide for the block execution logic.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by greged93**
This PR introduces an implementation of the `BlockExecutionStrategy` for Scroll. 
It also patches the `revm` crate to the [scroll fork](https://github.com/scroll-tech/revm/tree/scroll-evm-executor/reth/v50).

Resolves #5

### CODE REVIEW

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [inline crates/scroll/consensus/src/curie.rs:27]: Can we add a link to the verified contract on scroll scan?

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [inline crates/scroll/consensus/src/curie.rs:20]: Can we add some module documentation here that describes the migration of the oracle for the `Curie` hardfork and some insights into the fee mechanism or a reference to somewhere that describes it?

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [inline crates/scroll/consensus/src/curie.rs]: Lets add some doc comments to these constants.

File: Cargo.toml
Comment (frisitano) [inline Cargo.toml:678]: remark: Is there any computational overhead of using the forked `alloy-*` libraries that are required for sp1? If there is then I would propose that we feature gate the dependencies in the `revm` fork. I.e. have a `default` feature that uses the native libraries and an `sp1` feature that uses the patched libraries. I know we intend to migrate to the SDK pattern but I believe this will also be a consideration there. Can we create an issue in the `revm` to consider this.

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [inline crates/scroll/consensus/src/curie.rs]: Let's add more comprehensive documentation here that describes what changes are being made.

File: crates/scroll/consensus/src/lib.rs
Comment (frisitano) [inline crates/scroll/consensus/src/lib.rs]: nit: reverse order of `mod` and `pub use`

File: crates/scroll/evm/Cargo.toml
Comment (frisitano) [inline crates/scroll/evm/Cargo.toml:25]: question: is the `no_base_fee` feature used to disable the native ethereum base fee handling?

File: crates/scroll/evm/src/config.rs
Comment (frisitano) [inline crates/scroll/evm/src/config.rs:123]: Is it fine to use `default` is the `base_fee_per_gas` is not provided in the header or should we panic?

File: crates/scroll/evm/src/config.rs
Comment (frisitano) [inline crates/scroll/evm/src/config.rs]: question: Is the beneficiary in the header populated with the fee vault?

File: crates/scroll/evm/src/config.rs
Comment (frisitano) [inline crates/scroll/evm/src/config.rs:152]: Isn't this inconsistent with the logic on lines 115 - 121 in which we have different functionality for pre and post merge?

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [inline crates/scroll/evm/src/execute.rs]: Do we still need to rlp encode the transaction if it is an L1 message or can we set `evm.context.evm.env.tx.scroll.rlp_bytes = None`? Where is it used in the `evm`?

**Review by frisitano — COMMENTED** [review]
Left some initial comments and questions in line. Will review more thoroughly once we integrate the chain spec.

File: Cargo.toml
Comment (greged93) [reply Cargo.toml:678]: I checked, the only change in the alloy-eips fork is a `rkyv` feature. If you prefer I can just remove the patch though.

File: crates/scroll/evm/Cargo.toml
Comment (greged93) [reply crates/scroll/evm/Cargo.toml:25]: the `optional_no_base_fee` is used for the l1 messages which don't incur additional costs on the L2.

File: crates/scroll/evm/src/config.rs
Comment (greged93) [reply crates/scroll/evm/src/config.rs:123]: I think this is ok, this is taken from the [default implementation](https://github.com/paradigmxyz/reth/blob/2179301590b3dec0598910f553042ac078cfe1ac/crates/evm/src/lib.rs#L160-L178) for the trait. I initially wanted to do something like this:
```rust
    ConfigureEvmEnv::fill_block_env(...);
    if let Some(vault_address) = self.chain_spec.fee_vault_address {
            block_env.coinbase = vault_address;
    }
```
but this isn't possible because `fill_block_env` takes `&self`

File: Cargo.toml
Comment (frisitano) [reply Cargo.toml:678]: Unless there is significant performance justification to remove it then I think the benefit of maintaining consistency with the prover probably wins out as we need to balance the risk of there being some discrepancy that results in an unprovable block (even if the probability is low). What are your thoughts?

File: crates/scroll/evm/src/config.rs
Comment (greged93) [reply crates/scroll/evm/src/config.rs]: Unfortunately not: https://github.com/scroll-tech/reth/issues/5#issuecomment-2500852817

File: crates/scroll/evm/src/config.rs
Comment (greged93) [reply crates/scroll/evm/src/config.rs:152]: looking at the [documentation](https://github.com/paradigmxyz/reth/blob/2179301590b3dec0598910f553042ac078cfe1ac/crates/evm/src/lib.rs#L220-L232) for the `NextBlockEnvAttributes`, it looks like `next_cfg_and_block_env` and `fill_block_env` are complementary.

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: Yes we need to fill those otherwise we will [error](https://github.com/scroll-tech/revm/blob/scroll-evm-executor/v49/crates/revm/src/scroll/handler_register.rs#L58-L62) during execution.

File: Cargo.toml
Comment (greged93) [reply Cargo.toml:678]: I feel the same, will take the habit of examining changes of the patch every time we update it.

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [reply crates/scroll/evm/src/execute.rs]: But isn't that exclusively for the branch in which [`is_l1_message == false`](https://github.com/scroll-tech/revm/blob/91708073cefa4d22767d5a69a6636e68c09084e4/crates/revm/src/scroll/handler_register.rs#L53-L62). I think the branch for [`is_l1_messsage == true`](https://github.com/scroll-tech/revm/blob/91708073cefa4d22767d5a69a6636e68c09084e4/crates/revm/src/scroll/handler_register.rs#L81-L92) we don't use `.rlp_bytes` however I could be missing something.

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [inline crates/scroll/consensus/src/curie.rs:7]: We have support for `EIP-1559` and `EIP-2930`? The docs suggest we don't but they may be outdated https://docs.scroll.io/en/technology/chain/transactions/#transaction-type.

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: you might be right, will check the the fork and update if needed

**Review by frisitano — COMMENTED** [review]
One comment left inline.

File: crates/scroll/consensus/src/curie.rs
Comment (greged93) [reply crates/scroll/consensus/src/curie.rs:7]: As far as I can see [here](https://arc.net/l/quote/ucmdvtxg) yes

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [inline crates/scroll/evm/src/execute.rs]: maybe I also need to add 1559 and 2930 before Curie? I'm actually not sure if this belongs here or in the `validate_block_pre_execution` method of the Consensus trait?

File: crates/scroll/consensus/src/curie.rs
Comment (frisitano) [reply crates/scroll/consensus/src/curie.rs:7]: Opened an issue for the docs https://github.com/scroll-tech/scroll-documentation/issues/407

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [reply crates/scroll/evm/src/execute.rs:121]: Is there any handling of transaction types in revm itself?

Comment (frisitano) [reply crates/scroll/evm/src/execute.rs]: I'm not sure at this point. Can you create an issue `Milestone 1 Open Questions` (and add it to reth project and milestone 1) and add this as an open question? Let's keep that as a running list of things we need to figure out.

File: crates/scroll/evm/src/config.rs
Comment (greged93) [reply crates/scroll/evm/src/config.rs:152]: actually also looking at the [documentation](https://github.com/paradigmxyz/reth/blob/2179301590b3dec0598910f553042ac078cfe1ac/crates/evm/src/lib.rs#L208-L212) for the trait itself, the call to this method is intended for block building only post merge.

File: crates/scroll/chainspec/src/genesis.rs
Comment (frisitano) [inline crates/scroll/chainspec/src/genesis.rs]: nit: Can we define these values as constants with doc comments?

File: crates/scroll/evm/src/config.rs
Comment (frisitano) [inline crates/scroll/evm/src/config.rs]: Should we set the `block_env.coinbase` using:

```rust
        if let Some(vault_address) = self.scroll_config.fee_vault_address {
            block_env.coinbase = vault_address;
        } else {
            block_env.coinbase = header.beneficiary;
        }
```

This should prevent the additional assignment on line 108.

File: crates/scroll/evm/src/error.rs
Comment (frisitano) [inline crates/scroll/evm/src/error.rs]: Do we want to capture the database error inside of Curie i.e:

```rust
/// Scroll fork error.
#[derive(Debug, Display)]
pub enum ForkError {
    /// Error occurred at Curie fork.
    Curie(String),
}
```

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [inline crates/scroll/evm/src/execute.rs]: Wondering if there is any gas metering for this in l2geth?

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [reply crates/scroll/evm/src/execute.rs]: If we are going to keep this here then maybe we should add similar for EIP-7702 transactions.

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [inline crates/scroll/evm/src/execute.rs]: Given that we set this in lines 130 - 134 maybe it's fine to use `expect("this field is set on lines 130 - 134")` instead of `ok_or_else(..)`. What do you think?

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [inline crates/scroll/evm/src/execute.rs]: Do we have this constant set anywhere (not in test code) that we can leverage?

File: crates/scroll/evm/src/execute.rs
Comment (frisitano) [inline crates/scroll/evm/src/execute.rs]: Should we add coverage for EIP-2930, EIP-1559 and EIP-7702 as well?

**Review by frisitano — CHANGES_REQUESTED** [review]
Looks good! Left a few comments inline.

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: I did start out with an `expect`, but decided to switch, the reasoning being that we provide a `&mut Evm` to `revm` for the call to transact, so (theoretically) the `revm` implementation could decide to `take` the RLP bytes. If you think it's ok to switch to `expect` I'm happy to update.

Comment (frisitano) [reply crates/scroll/evm/src/execute.rs]: Yes, I think it's fine to use `expect` here. `revm` is deterministic and we know based on the logic that this will never happen.

Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: same for the `l1_block_info` then? this should always be set in the `load_accounts` hook of the handler.

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: I don't think so, the state just gets updated without any gas being accounted in the block.
https://github.com/scroll-tech/go-ethereum/blob/develop/consensus/misc/curie.go#L12-L23

File: crates/scroll/evm/src/execute.rs
Comment (greged93) [reply crates/scroll/evm/src/execute.rs]: unfortunately not, I thought maybe in the `reth-scroll-chainspec` but it doesn't look like it.

File: crates/scroll/chainspec/src/constants.rs
Comment (frisitano) [inline crates/scroll/chainspec/src/constants.rs]: Let's make a constant for this.

File: crates/scroll/chainspec/src/constants.rs
Comment (frisitano) [inline crates/scroll/chainspec/src/constants.rs]: Can you take the chain id from `alloy-chains`? Similarly fro scroll sepolia below.

File: crates/scroll/evm/src/config.rs
Comment (frisitano) [inline crates/scroll/evm/src/config.rs:22]: question: Should this be `Arc< ScrollChainConfig>`?

**Review by frisitano — DISMISSED** [review]
Looks good! Left minor nits / question inline. After that we can merge.

File: crates/scroll/evm/src/config.rs
Comment (greged93) [reply crates/scroll/evm/src/config.rs:22]: I think the idea of the `Arc` for the `ScrollChainSpec` is that cloning it might incur a high cost if the genesis is large. In the case of the `ScrollChainConfig`, I think the cost for cloning is a lot more limited so I decided to keep it as is.

**Review by frisitano — APPROVED** [review]
Looks great, thanks! Lets wait for CI to finish and then we can merge.

**Pull request merged**
